File: Python/Medium/backtracking/combination_sum.py
class Solution:
    def combinationSum(self, candidates, target):
        result = []

        def backtrack(start, current_sum, current):
            if current_sum > target:
                return
            elif current_sum == target:
                result.append(current.copy())
                return
            for i in range(start, len(candidates)):
                current.append(candidates[i])

                backtrack(start, current_sum+candidates[i], current)

                current.pop()

        backtrack(0,0,[])
        return result 

File: JavaScript/Medium/backtracking/combination_sum.js
/**
 * @param {number []} candidates
 * @param {number} target
 * @return {number [][]} 
 */

var combinationSum = (candidates, target) =>{
    let result = []
    var backtrack = (start, current_sum, current) => {
        if (current_sum > target){
            return
        }
        if (current_sum === target){
            result.push([...current])
            return
        }
        for (let i = 0; start<candidates.length; i++){
            current.push(candidates[i])

            backtrack(i, current_sum+candidates[i], current)

            current.pop()
        }
    }
    backtrack(0,0,[])
    return result
}